### PR TITLE
Aggressive warm up duration fix and sync client warm up fix

### PR DESCRIPTION
**Pull request by jeet1995**
This PR fixes the following bugs:
- Previously the building of a `CosmosClient` or `CosmosAsyncClient` would block until aggressive warm up duration elapses. This should not be the case when warm up can complete before this duration.
- The `CosmosClient` opened connections defensively even during the warmup phase. Defensive connection establishment should only happen after the client is built.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 1]
```diff
@@ -27,16 +31,9 @@
 #### Other Changes
 * Added coverage for `ChangeFeedProcessor` in `Latest Version` change feed mode to read change feed from a custom start time for multi-write accounts. - See[PR 36257](https://github.com/Azure/azure-sdk-for-java/pull/36257)
 
-### 4.48.2 (2023-08-25)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:30]: why this change

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 2]
```diff
@@ -89,25 +89,14 @@ there are non-existent document IDs also passed through the API - See [PR 35513]
   JVM configuration - `COSMOS.MIN_CONNECTION_POOL_SIZE_PER_ENDPOINT` - See [PR 34859](https://github.com/Azure/azure-sdk-for-java/pull/34859)
 * Extending maximum retry delay in `SessionTokenMismatchRetryPolicy`. - See [PR 35360](https://github.com/Azure/azure-sdk-for-java/pull/35360)
 
-### 4.45.2-hotfix (2023-09-18)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:92]: same here

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java [hunk 3]
```diff
@@ -614,7 +614,9 @@ void openConnectionsAndInitCaches(Duration aggressiveWarmupDuration) {
     // with a sink and block on the wrapping flux for the specified duration
     private Flux<Void> wrapSourceFluxAndSoftCompleteAfterTimeout(Flux<Void> source, Duration timeout) {
         return Flux.<Void>create(sink -> {
-                    source.subscribe(t -> sink.next(t));
+                    source
+                        .doOnComplete(() -> sink.complete())
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java:618]: would be doFinally safer here, for case where the source failed

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java [hunk 4]
```diff
@@ -636,14 +742,50 @@ private Object[] proactiveContainerInitConfigs() {
...
+            // the client build should not block until this aggressive warm up duration time
+            new ProactiveConnectionManagementTestConfig()
+                .withPreferredRegions(preferredRegions)
+                .withProactiveConnectionRegionsCount(2)
+                .withContainerCount(3)
+                .withMinConnectionPoolSizePerEndpoint(5)
+                .withAggressiveWarmupDuration(Duration.ofMinutes(1000))
+                .withIsSystemPropertySetBeforeDirectConnectionConfig(false)
+                .withIsSyncClient(false),
+            new ProactiveConnectionManagementTestConfig()
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java:765]: since we introduced faultInjection on gateway, can we also add test cases where some errors during the warm up stage - for example collection read timed out etc. - can be in a different PR

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 1 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:30]: Faulty merge conflict resolution. Fixing this in next iteration.

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:92]: Faulty merge conflict resolution. Fixing this in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java (hunk 3 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java:618]: Good catch. Fixed in next iteration.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java (hunk 4 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java:765]: Tracking the work item here - https://github.com/Azure/azure-sdk-for-java/issues/33088

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,9 @@
 
 #### Bugs Fixed
 * Fixed staleness issue of `COSMOS.MIN_CONNECTION_POOL_SIZE_PER_ENDPOINT` system property - See [PR 36599](https://github.com/Azure/azure-sdk-for-java/pull/36599).
+* Fixed to not stall building of a client for the entire aggressive warm up duration but instead until containers
+in the warmup flow have been warmed up or until aggressive warm up duration has been reached whichever is lesser. - See [PR 36889](https://github.com/Azure/azure-sdk-for-java/pull/36889)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:13]: Fixed an issue where cosmos client was getting blocked for the entire aggressive warmup duration even when all the connections have been opened already - See [PR 36889](https://github.com/Azure/azure-sdk-for-java/pull/36889)

**Review by kushagraThapar — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 5 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:13]: better to keep the changelog wordings simple :)

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 5 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md]: Fixed in next iteration.
